File: ambeth/Ambeth.Security/ambeth/privilege/model/impl/PropertyPrivilegeImpl.cs
using De.Osthus.Ambeth.Privilege.Transfer;
using De.Osthus.Ambeth.Util;
using System;
using System.Text;

namespace De.Osthus.Ambeth.Privilege.Model.Impl
{
    public sealed class PropertyPrivilegeImpl : IPropertyPrivilege, IPrintable, IImmutableType
    {
        public static readonly IPropertyPrivilege[] EMPTY_PROPERTY_PRIVILEGES = new IPropertyPrivilege[0];

        private static readonly PropertyPrivilegeImpl[] array = new PropertyPrivilegeImpl[ArraySizeForIndex()];

        static PropertyPrivilegeImpl()
        {
            Put1();
        }

        private static void Put1()
        {
            Put2(true);
            Put2(false);
        }

        private static void Put2(bool create)
        {
            Put3(create, true);
            Put3(create, false);
        }

        private static void Put3(bool create, bool read)
        {
            Put4(create, read, true);
            Put4(create, read, false);
        }

        private static void Put4(bool create, bool read, bool update)
        {
            Put(create, read, update, true);
            Put(create, read, update, false);
        }

        public static int ArraySizeForIndex()
        {
            return 1 << 4;
        }

        public static int CalcIndex(bool create, bool read, bool update, bool delete)
        {
            return AbstractPrivilege.ToBitValue(create, 0) + AbstractPrivilege.ToBitValue(read, 1) + AbstractPrivilege.ToBitValue(update, 2)
                    + AbstractPrivilege.ToBitValue(delete, 3);
        }

        private static void Put(bool create, bool read, bool update, bool delete)
        {
            int index = CalcIndex(create, read, update, delete);
            array[index] = new PropertyPrivilegeImpl(create, read, update, delete);
        }

        public static IPropertyPrivilege Create(bool create, bool read, bool update, bool delete)
        {
            int index = CalcIndex(create, read, update, delete);
            return array[index];
        }
        
        public static IPropertyPrivilege CreateFrom(IPrivilege privilegeAsTemplate)
        {
            return Create(privilegeAsTemplate.CreateAllowed, privilegeAsTemplate.ReadAllowed, privilegeAsTemplate.UpdateAllowed,
                    privilegeAsTemplate.DeleteAllowed);
        }

        public static IPropertyPrivilege CreateFrom(IPrivilegeOfService privilegeOfService)
        {
            return Create(privilegeOfService.CreateAllowed, privilegeOfService.ReadAllowed, privilegeOfService.UpdateAllowed,
                    privilegeOfService.DeleteAllowed);
        }

        public static IPropertyPrivilege CreateFrom(IPropertyPrivilegeOfService propertyPrivilegeResult)
        {
            return Create(propertyPrivilegeResult.CreateAllowed, propertyPrivilegeResult.ReadAllowed, propertyPrivilegeResult.UpdateAllowed,
                    propertyPrivilegeResult.DeleteAllowed);
        }

        private bool create;
        private bool read;
        private bool update;
        private bool delete;

        private PropertyPrivilegeImpl(bool create, bool read, bool update, bool delete)
        {
            this.create = create;
            this.read = read;
            this.update = update;
            this.delete = delete;
        }

        public bool CreateAllowed
        {
            get { return create; }
        }

        public bool ReadAllowed
        {
            get { return read; }
        }

        public bool UpdateAllowed
        {
            get { return update; }
        }

        public bool DeleteAllowed
        {
            get { return delete; }
        }
        
        public override String ToString()
        {
            StringBuilder sb = new StringBuilder();
            ToString(sb);
            return sb.ToString();
        }

        public void ToString(StringBuilder sb)
        {
            sb.Append(AbstractPrivilege.upperOrLower(CreateAllowed, 'c'));
            sb.Append(AbstractPrivilege.upperOrLower(ReadAllowed, 'r'));
            sb.Append(AbstractPrivilege.upperOrLower(UpdateAllowed, 'u'));
            sb.Append(AbstractPrivilege.upperOrLower(DeleteAllowed, 'd'));
        }
    }
}

File: jambeth/jambeth-security/src/main/java/de/osthus/ambeth/privilege/model/impl/PropertyPrivilegeImpl.java
package de.osthus.ambeth.privilege.model.impl;

import java.io.ObjectStreamException;
import java.io.Serializable;

import de.osthus.ambeth.privilege.model.IPrivilege;
import de.osthus.ambeth.privilege.model.IPropertyPrivilege;
import de.osthus.ambeth.privilege.transfer.IPrivilegeOfService;
import de.osthus.ambeth.privilege.transfer.IPropertyPrivilegeOfService;
import de.osthus.ambeth.util.IImmutableType;
import de.osthus.ambeth.util.IPrintable;

public final class PropertyPrivilegeImpl implements IPropertyPrivilege, IPrintable, IImmutableType, Serializable
{
	private static final long serialVersionUID = 3959654761266019980L;

	public static final IPropertyPrivilege[] EMPTY_PROPERTY_PRIVILEGES = new IPropertyPrivilege[0];

	private static final PropertyPrivilegeImpl[] array = new PropertyPrivilegeImpl[arraySizeForIndex()];

	static
	{
		put1();
	}

	private static void put1()
	{
		put2(true);
		put2(false);
	}

	private static void put2(boolean create)
	{
		put3(create, true);
		put3(create, false);
	}

	private static void put3(boolean create, boolean read)
	{
		put4(create, read, true);
		put4(create, read, false);
	}

	private static void put4(boolean create, boolean read, boolean update)
	{
		put(create, read, update, true);
		put(create, read, update, false);
	}

	public static int arraySizeForIndex()
	{
		return 1 << 4;
	}

	public static int calcIndex(boolean create, boolean read, boolean update, boolean delete)
	{
		return AbstractPrivilege.toBitValue(create, 0) + AbstractPrivilege.toBitValue(read, 1) + AbstractPrivilege.toBitValue(update, 2)
				+ AbstractPrivilege.toBitValue(delete, 3);
	}

	private static void put(boolean create, boolean read, boolean update, boolean delete)
	{
		int index = calcIndex(create, read, update, delete);
		array[index] = new PropertyPrivilegeImpl(create, read, update, delete);
	}

	public static IPropertyPrivilege create(boolean create, boolean read, boolean update, boolean delete)
	{
		int index = calcIndex(create, read, update, delete);
		return array[index];
	}

	public static IPropertyPrivilege createFrom(IPrivilege privilegeAsTemplate)
	{
		return create(privilegeAsTemplate.isCreateAllowed(), privilegeAsTemplate.isReadAllowed(), privilegeAsTemplate.isUpdateAllowed(),
				privilegeAsTemplate.isDeleteAllowed());
	}

	public static IPropertyPrivilege createFrom(IPrivilegeOfService privilegeOfService)
	{
		return create(privilegeOfService.isCreateAllowed(), privilegeOfService.isReadAllowed(), privilegeOfService.isUpdateAllowed(),
				privilegeOfService.isDeleteAllowed());
	}

	public static IPropertyPrivilege createFrom(IPropertyPrivilegeOfService propertyPrivilegeResult)
	{
		return create(propertyPrivilegeResult.isCreateAllowed(), propertyPrivilegeResult.isReadAllowed(), propertyPrivilegeResult.isUpdateAllowed(),
				propertyPrivilegeResult.isDeleteAllowed());
	}

	private final boolean create;
	private final boolean read;
	private final boolean update;
	private final boolean delete;

	private PropertyPrivilegeImpl(boolean create, boolean read, boolean update, boolean delete)
	{
		this.create = create;
		this.read = read;
		this.update = update;
		this.delete = delete;
	}

	@Override
	public boolean isCreateAllowed()
	{
		return create;
	}

	@Override
	public boolean isReadAllowed()
	{
		return read;
	}

	@Override
	public boolean isUpdateAllowed()
	{
		return update;
	}

	@Override
	public boolean isDeleteAllowed()
	{
		return delete;
	}

	@Override
	public final String toString()
	{
		StringBuilder sb = new StringBuilder();
		toString(sb);
		return sb.toString();
	}

	@Override
	public void toString(StringBuilder sb)
	{
		sb.append(AbstractPrivilege.upperOrLower(isCreateAllowed(), 'c'));
		sb.append(AbstractPrivilege.upperOrLower(isReadAllowed(), 'r'));
		sb.append(AbstractPrivilege.upperOrLower(isUpdateAllowed(), 'u'));
		sb.append(AbstractPrivilege.upperOrLower(isDeleteAllowed(), 'd'));
	}

	private Object readResolve() throws ObjectStreamException
	{
		return create(create, read, update, delete);
	}
}
